ha_mroonga.cpp

```diff
@@ -14954,56 +14954,123 @@ mrn_io_and_cpu_cost ha_mroonga::scan_time()
+    res = wrap_handler->MRN_HANDLER_KEYREAD_TIME(share->wrap_key_nr[index],
+                                                 ranges,
+                                                 rows,
+                                                 blocks);
     MRN_SET_BASE_SHARE_KEY(share, table->s);
     MRN_SET_BASE_TABLE_KEY(this, table);
   } else {
     MRN_SET_WRAP_SHARE_KEY(share, table->s);
     MRN_SET_WRAP_TABLE_KEY(this, table);
-    res = wrap_handler->read_time(index, ranges, rows);
+    res = wrap_handler->MRN_HANDLER_KEYREAD_TIME(index, ranges, rows, blocks);
     MRN_SET_BASE_SHARE_KEY(share, table->s);
     MRN_SET_BASE_TABLE_KEY(this, table);
   }
   DBUG_RETURN(res);
 }
 #endif
 
-double ha_mroonga::storage_read_time(uint index, uint ranges, ha_rows rows)
+mrn_io_and_cpu_cost ha_mroonga::storage_read_time(uint index,
+                                                  uint ranges,
+                                                  ha_rows rows,
+                                                  ulonglong blocks)
 {
   MRN_DBUG_ENTER_METHOD();
-  double time = handler::read_time(index, ranges, rows);
+  mrn_io_and_cpu_cost time =
+    handler::MRN_HANDLER_KEYREAD_TIME(index, ranges, rows, blocks);
   DBUG_RETURN(time);
 }
 
+#ifdef MRN_HANDLER_USE_KEYREAD_TIME
+IO_AND_CPU_COST ha_mroonga::keyread_time(uint index,
+                                         uint ranges,
+                                         ha_rows rows,
+                                         ulonglong blocks)
+#endif
+#ifdef MRN_HANDLER_HAVE_READ_TIME
 double ha_mroonga::read_time(uint index, uint ranges, ha_rows rows)
+#endif
 {
   MRN_DBUG_ENTER_METHOD();
-  double time;
+  mrn_io_and_cpu_cost time;
 #ifdef MRN_ENABLE_WRAPPER_MODE
   if (share->wrapper_mode)
   {
-    time = wrapper_read_time(index, ranges, rows);
+    time = wrapper_read_time(index, ranges, rows
+#  ifdef MRN_HANDLER_USE_KEYREAD_TIME
+                             , blocks
+#  endif
+                             );
   } else {
 #endif
-    time = storage_read_time(index, ranges, rows);
+    time = storage_read_time(index, ranges, rows
+#ifdef MRN_HANDLER_USE_KEYREAD_TIME
+                             , blocks
+#endif
+                             );
 #ifdef MRN_ENABLE_WRAPPER_MODE
   }
 #endif
   DBUG_RETURN(time);
 }
 
+#ifdef MRN_HANDLER_USE_KEYREAD_TIME
+#  ifdef MRN_ENABLE_WRAPPER_MODE
+IO_AND_CPU_COST ha_mroonga::wrapper_rnd_pos_time(ha_rows rows)
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST res;
+  MRN_SET_WRAP_SHARE_KEY(share, table->s);
+  MRN_SET_WRAP_TABLE_KEY(this, table);
+  res = wrap_handler->rnd_pos_time(rows);
+  MRN_SET_BASE_SHARE_KEY(share, table->s);
+  MRN_SET_BASE_TABLE_KEY(this, table);
+  DBUG_RETURN(res);
+}
+#  endif
+
+IO_AND_CPU_COST ha_mroonga::storage_rnd_pos_time(ha_rows rows)
+{
+  MRN_DBUG_ENTER_METHOD();
+  IO_AND_CPU_COST res;
+  res = handler::rnd_pos_time(rows);
+  DBUG_RETURN(res);
+}
+
+IO_AND_CPU_COST ha_mroonga::rnd_pos_time(ha_rows rows)
```

Comment:
We don't need to override `rnd_pos_time()` when wrapper mode isn't enabled because we use the default `handler::rnd_pos_time()` implementation for storage mode, right?